gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/KillDagProc.java

```diff
@@ -0,0 +1,114 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration.proc;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import com.google.common.collect.Maps;
+
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.orchestration.DagManagementStateStore;
+import org.apache.gobblin.service.modules.orchestration.DagManagerUtils;
+import org.apache.gobblin.service.modules.orchestration.NewDagManager;
+import org.apache.gobblin.service.modules.orchestration.TimingEventUtils;
+import org.apache.gobblin.service.modules.orchestration.task.DagTask;
+import org.apache.gobblin.service.modules.orchestration.task.KillDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+
+import static org.apache.gobblin.service.ExecutionStatus.CANCELLED;
+
+/**
+ * An implementation of {@link DagProc} for killing {@link DagTask}.
+ */
+@Slf4j
+@Alpha
+public final class KillDagProc extends DagProc<KillDagProc.CancelEntity, Dag<JobExecutionPlan>> {
+
+  // should dag task be a part of dag proc?
+  private final KillDagTask killDagTask;
+
+  public KillDagProc(KillDagTask killDagTask, NewDagManager dagManager) {
+    super(dagManager);
+    this.killDagTask = killDagTask;
+  }
+
+  protected CancelEntity initialize(DagManagementStateStore dagManagementStateStore) throws IOException {
+    Dag<JobExecutionPlan> dagToCancel = dagManagementStateStore.getDag(this.killDagTask.getDagId().toString());
+    List<Dag.DagNode<JobExecutionPlan>> dagNodesToCancel = new ArrayList<>(dagManagementStateStore.getJobs(this.killDagTask.getDagId().toString()));
+    return new CancelEntity(dagToCancel, dagNodesToCancel);
+  }
+
+  @Override
+  public Dag<JobExecutionPlan> act(DagManagementStateStore dagManagementStateStore, CancelEntity cancelEntity) throws IOException {
+    if (cancelEntity.dagToCancel == null || cancelEntity.dagNodesToCancel.isEmpty()) {
+      log.warn("No dag with id " + this.killDagTask.getDagId() + " found to kill");
+      return null;
```

Comment:
let's make it clear where null can be returned and make decision between null and Optional parameters. Kip and I have raised it in the past that null does not make it obvious to caller that the object may not be present and can cause NPE